### PR TITLE
Allow broadcasting in Merge layer

### PR DESCRIPTION
**Pull request by farizrahman4u**
```python
from keras.layers import *
from keras.models import *

a = Input((10, 5))  # batch_size, 10, 5
b = Input((5,))  # batch_size, 5
c = add([a, b])  # batch_size, 10, 5
model = Model([a, b], c)
```

### CODE REVIEW

**Comment by farizrahman4u** [comment]
ping @fchollet

File: keras/layers/merge.py [hunk 1]
```diff
@@ -18,6 +18,31 @@ def __init__(self, **kwargs):
...
+        output_shape = list(shape1[:-len(shape2)])
+        for i, j in zip(shape1[-len(shape2):], shape2):
+            if i is None or j is None:
+                output_shape.append(None)
+            elif i == 1:
+                output_shape.append(j)
+            elif j == 1:
+                output_shape.append(i)
+            else:
+                assert i == j, 'Operands could not be broadcast together with shapes ' + str(shape1) + ' ' + str(shape2)
```
Comment (fchollet) [inline keras/layers/merge.py]: Line is too long. Raise a ValueError insert of using an assert.

File: keras/layers/merge.py [hunk 2]
```diff
@@ -18,6 +18,31 @@ def __init__(self, **kwargs):
     def _merge_function(self, inputs):
         raise NotImplementedError
 
+    def _compute_elemwise_op_output_shape(self, shape1, shape2):
+        """check if shapes of 2 tensors are compatible for element wise operations,
```
Comment (fchollet) [inline keras/layers/merge.py]: Format the docstring in the style of other docstrings in the codebase. In particular, needs `# Arguments`, `# Returns` and `# Raises`

File: keras/layers/merge.py [hunk 3]
```diff
@@ -27,19 +52,83 @@ def build(self, input_shape):
...
-        if None in input_shapes_set:
-            input_shapes_set.remove(None)
-        if len(input_shapes_set) > 1:
-            raise ValueError('Only tensors of same shape can '
-                             'be merged by layer' + self.name +
-                             ' Got input shapes: %s' % input_shape)
+        batch_sizes = [s[0] for s in input_shape if s is not None]
+        batch_sizes = set(batch_sizes)
+        batch_sizes -= set([None])
+        assert len(batch_sizes) <= 1, 'Can not merge tensors with different batch sizes. Got tensors with shapes : ' + str(input_shape)
```
Comment (fchollet) [inline keras/layers/merge.py]: Line is too long. Raise a ValueError insert of using an assert.

**Comment by farizrahman4u** [comment]
> it seems quite complicated 

It is. Though operator broadcasting is supported implicitly by both backends, they do not take batch size into account. For example, in TH/TF you can add 2 tensors with shapes `(10, 5)` and `(5,)`  but you can not do it for `(32, 10, 5)` and `(32, 5)` where 32 is the batch size. So you have to do a transpose on the inputs and make them `(10, 5, 32)` and `(5, 32)`.. now they are broadcast ready. After the operation, we have to transpose the result back.

To make things more complicated, it is possible for TF tensors to have an undefined ndim, which makes the transposing more complex. Also TF doesn't support negative indexing in some places.

**Comment by farizrahman4u** [comment]
@fchollet Checkout unit tests.

**Comment by farizrahman4u** [comment]
ping @fchollet

File: keras/layers/merge.py [hunk 4]
```diff
@@ -27,19 +66,86 @@ def build(self, input_shape):
...
+            else:
+                shape = input_shape[i][1:]
+            output_shape = self._compute_elemwise_op_output_shape(output_shape, shape)
+        batch_sizes = [s[0] for s in input_shape if s is not None]
+        batch_sizes = set(batch_sizes)
+        batch_sizes -= set([None])
+        if len(batch_sizes) == 1:
+            output_shape = (batch_sizes[0],) + output_shape
+        else:
+                output_shape = (None,) + output_shape
```
Comment (fchollet) [inline keras/layers/merge.py]: Indentation issue.

File: keras/layers/merge.py [hunk 5]
```diff
@@ -27,19 +66,86 @@ def build(self, input_shape):
...
+        else:
+            output_shape = input_shape[0][1:]
+        for i in range(1, len(input_shape)):
+            if input_shape[i] is None:
+                shape = None
+            else:
+                shape = input_shape[i][1:]
+            output_shape = self._compute_elemwise_op_output_shape(output_shape, shape)
+        if None not in input_shape and len(set(map(len, input_shape))) == 1:
+            self._permute_required = False
```
Comment (fchollet) [inline keras/layers/merge.py]: Please add a comment explaining the purpose of `_permute_required`, it's not obvious from the name.

File: keras/layers/merge.py [hunk 6]
```diff
@@ -27,19 +66,86 @@ def build(self, input_shape):
...
+        else:
+            self._permute_required = True
 
     def call(self, inputs):
-        return self._merge_function(inputs)
+        if self._permute_required:
```
Comment (fchollet) [inline keras/layers/merge.py]: Please add a comment explaining what goes on below.

File: keras/layers/merge.py [hunk 7]
```diff
@@ -27,19 +66,86 @@ def build(self, input_shape):
...
+            reshaped_inputs = []
+            reshaped = False
+            for x in inputs:
+                x_ndim = K.ndim(x)
+                if x_ndim is None:
```
Comment (fchollet) [inline keras/layers/merge.py]: This case is sufficiently irregular that we could simply raise a ValueError and required `ndim` to be defined. Seems like it would simplify the code quite a bit.

File: keras/layers/merge.py (hunk 2 above)
Comment (fchollet) [inline keras/layers/merge.py]: For style purposes, the first line in the docstring should fit in one line (here you have 2), should be capitalized and end in a dot. This is automatically enforced in tf-keras, so we have to make the codebase compliant.

File: keras/layers/merge.py (hunk 7 above)
Comment (farizrahman4u) [inline keras/layers/merge.py]: This case pops up in tests. Doing so will cause tests to fail.

**Comment by fchollet** [comment]
Looks good to me. Since these are fairly involved changed, anyone else beside me would like to have a second look?

**Comment by angeloskath** [comment]
@farizrahman4u I think this is more complicated than needed. For instance there is no need for transposition when broadcasting for the case you mentioned above.

The rules are simple:
- Discard (I mean do not worry about it) the batch dimension which is checked in other places in Keras
- Get all shapes (after the batch dim) as None or ints
- Prepend all shapes with ones so that they have the same number of dimensions
- Check for broadcastable shapes
    1. 1 + x --> x
    2. x + x --> x
    3. Anything else is not allowed

**Comment by farizrahman4u** [comment]
@angeloskath I dont get you. Are you talking about shape inference or or the actual computation itself? We dont pass the batch size to the shape inference function. And transposing is required to broadcast tensors with different ndims.

**Comment by angeloskath** [comment]
Ok, I 've been reading more carefully through your tests and code to see why so much code is needed for this. I have a couple of suggestions which may or may not be good.

Your implementation tries to always build the layer if there is even a probability (not certainty) that the operation can be carried out. The previous implementation was too restrictive, this is too permissive. For instance in your tests, what happens if someone changes line 142 to `x1 = np.random.random((2, 4, 6))`? The operation breaks inside tensorflow with a pretty big traceback. I believe that **we should not treat** `(None, None)` with `(None, None)` to be broadcastable. The same goes for missing number of dimensions. _On the other hand `(1,)` and `(None,)` is broadcastable with output `(None,)`_

Considering the above, the code can be simplified immensely. Afterwards concerning the transposition, I believe that you should be expanding all the dimensions in `axis=1` until all inputs have the same number of dimensions and then simply call the merge function. One could even argue that even different number of dimensions could be disallowed, we are simply delegating the responsibility of properly reshaping the inputs for broadcasting to the caller.

I hope I am not coming on too harsh, it's great code but I believe it tries to do too much.

**Comment by farizrahman4u** [comment]
It is possible to break a sufficiently complex model with `None` in input shape by providing a numpy arrays with unexpected shapes irrespective of this PR. The main use case of this feature is to add/mul 2D tensors to 3D tensors. So disallowing tensors of different ndims makes little to no sense. And your method of expanding dims instead of transposing won't work. (Please provide code snippet if it does.)

**Comment by angeloskath** [comment]
If @fchollet thinks that the Merge layer should account for tensors that do not provide `ndim` then lgtm. Otherwise removing that would simplify the code a lot.

**Comment by fchollet** [comment]
I was also wondering about that, since I don't think that case would normally arise with Keras tensors. But hey, it does make the layer more general for use with arbitrary TF tensors.
